Add `Client_main.find_master'`

NicolasT opened the pull request:
This patch adds `Client_main.find_master'`, which should be a more resilient master lookup algorithm.

It also reworks `Client_main.find_master` to use the new `find_master'`, but keep the behaviour of the original action regarding error handling.

NicolasT committed:
Add `Client_main.find_master'`

This patch adds `Client_main.find_master'`, which should be a more
resilient master lookup algorithm.

It also reworks `Client_main.find_master` to use the new `find_master'`,
but keep the behaviour of the original action regarding error handling.

NicolasT commented:
- [x] CI: `arakoon-git-launcher` 356
- [x] Attempt to connect to all IP addresses of a node concurrently?

domsj commented:
I think we should definitely try connecting on all ips of a node, and while it could be done concurrently doing it sequentially would already be an important improvement too (and actually be good enough for me).

Anyway, this definitely looks like an improvement so far.

NicolasT committed:
Implement concurrent connecting

NicolasT commented:
Implemented concurrent connecting.

NicolasT commented:
- [ ] `arakoon-git-launcher` 361

NicolasT committed:
Fix exception handling in `Client_main.with_connection'`

domsj commented on src/client/client_main.ml:
```diff
@@ -69,24 +69,68 @@ let with_connection ~tls sa do_it = match tls with
+            else begin
+              Lwt_mutex.with_lock l (fun () ->
+                Lwt_mvar.take result >>= function
+                  | Some _ -> Lwt.fail Done
```
- this can't actually happen
- if it did happen you would want to put the result back into the mvar

domsj commented on src/client/client_main.ml:
```diff
@@ -69,24 +69,68 @@ let with_connection ~tls sa do_it = match tls with
-        Lwt.fail exc)
+        with_connection ~tls addr (fun c ->
+          if Lwt_mutex.is_locked l
+            then Lwt.fail In_progress
```
Why not just `Lwt.return_unit`?

domsj commented on src/client/client_main.ml:
```diff
@@ -69,24 +69,68 @@ let with_connection ~tls sa do_it = match tls with
+
+exception In_progress
+exception Done
+exception Impossible
```
I don't think you need any of these...

NicolasT committed:
Address review comments

NicolasT committed:
Simplify & fix `Client_main.with_connection'`

NicolasT committed:
Rework `Client_main.with_connection'`

NicolasT committed:
Minor API enhancements

NicolasT commented on src/client/client_main.ml:
```diff
@@ -69,24 +69,109 @@ let with_connection ~tls sa do_it = match tls with
     f client
   in
-  with_connection ~tls sa do_it
+  with_connection' ~tls addrs do_it
```
This can raise other exceptions than the original implementation (`No_connection` instead of some `Unix_error`). Is that a problem?

NicolasT replied:
Fixed.

NicolasT committed:
Move `CountDownLatch` into a new `Lwt_extra` module

domsj commented on src/client/client_main.ml:
```diff
@@ -69,24 +69,109 @@ let with_connection ~tls sa do_it = match tls with
+    val await : t -> unit Lwt.t
+    val count_down : t -> unit Lwt.t
+end = struct
+    type t = { mvar : int Lwt_mvar.t
```
I don't understand why this is an mvar instead of a `int ref` ?
I think we could drop the Lwt_mutex.t below here too

domsj commented on src/client/client_main.ml:
```diff
@@ -69,24 +69,109 @@ let with_connection ~tls sa do_it = match tls with
+      (fun exn ->
+        CountDownLatch.count_down cd >>= fun () ->
+        Logger.warning_f_ ~exn "Failed to connect" >>= fun () ->
+        Lwt.fail exn)
```
Hmm, I expected all exceptions to be caught here, saved into some refcell, and then possible rethrown where you throw the No_connection now?
(That way with_connection would still throw the same exceptions, to answer your comment a bit below here)

domsj commented on src/tools/lwt_extra.ml:
```diff
@@ -0,0 +1,73 @@
+        Lwt_mutex.with_lock t.mutex (fun () ->
+            Lwt.protected (
+                Lwt_mvar.take t.mvar >>= fun c ->
+                let c' = max (c - 1) 0 in
```
Don't we want to throw an exception here when the count would go below 0?
Seems like an incorrect usage to me... (but I honestly haven't looked at APIs for countdown latches in other places, so not sure what the standard behavior there is)

NicolasT replied:
Not necessarily.

Say you got a networked system, consisting of 5 target nodes. In some algorithm, you want to send some message (and receive an acknowledgement) to at least 3 of these nodes, concurrently. Then you use a `CountDownLatch` of size 3, spawn 5 threads which send the message & wait for an answer, and once the ack has been received, decrease the latch.

Finally the caller blocks on the latch (& some timeout, or whatever).

It's perfectly possible and OK if the latch is decremented even though it's 0 already when more-than-3 messages are sent/ack'ed.

domsj replied:
Ok, sounds reasonable.
(Fwiw, apparently in Java it has the same semantics.)

NicolasT committed:
Pass through exceptions if applicable

Handling review comment
https://github.com/Incubaid/arakoon/pull/479#discussion_r14876029

domsj commented:
LG if CI is ok

NicolasT committed:
Merge origin/1.7 into find-master (using imerge)
